lib/AnalysisStructured/PtrAnalysis.cpp

```diff
@@ -0,0 +1,1128 @@
+    }
+
+    // Note that we want the knownPtrs to be indexed by block arg, but we
+    // only have index for now. Also, the state we record is the init
+    // arg, but want to to use newly created block arg. These block args
+    // are not created yet. We will translate this mapping later.
+    knownPtrsTmp.push_back(std::make_pair(i, state));
+    levelToBlockArgIndex[level].insert(i);
+  }
+
+  // Create a new scf::ForOp that uses updated init args and same loop body
+  auto newOp = builder.create<scf::ForOp>(
+      op.getLoc(), op.getLowerBound(), op.getUpperBound(), op.getStep(),
+      newInitArgs, [&](OpBuilder &b, Location loc, Value iv, ValueRange args) {
+        IRMapping cloneMap;
+        cloneMap.map(op.getInductionVar(), iv);
+        cloneMap.map(op.getInitArgs(), newInitArgs);
+        cloneMap.map(op.getRegionIterArgs(), args);
+
+        for (auto &bodyOp : op.getRegion().getOps()) {
+          b.clone(bodyOp, cloneMap);
+        }
+      });
+
+  // Convert the book-keeping data structure to use the correct key and value.
+  // Key is converted from init arg index to newly created block arg, and
+  // Value's PtrSState fields are converted from init arg to newly created block
+  // arg
+  int cnt = op.getRegionIterArgs().size();
+  for (auto [i, state] : knownPtrsTmp) {
+    for (auto it = state.offsets.begin(); it != state.offsets.end(); it++) {
+      *it = newOp.getRegionIterArgs()[cnt];
+      cnt++;
+    }
+
+    for (auto it = state.strides.begin(); it != state.strides.end(); it++) {
+      *it = newOp.getRegionIterArgs()[cnt];
+      cnt++;
+    }
+
+    if (state.getRank() == 0) {
+      assert(state.scalar);
+      state.scalar = newOp.getRegionIterArgs()[cnt];
+      cnt++;
+    }
+
+    // Record the PtrSState for this pointer
+    auto key = newOp.getRegionIterArgs()[i];
+    knownPtrs[key] = state;
+    initArgIndexMap[i] = state;
+
+    // Create a tts.make_tptr at the beginning of the loop body that correspond
+    // to this region iter arg. In case it is used by tt.load/tt.store in the
+    // loop body, this will make sure rewriteLoadOp/rewriteStoreOp can use the
+    // analysis result.
+    if (state.getRank() != 0) {
+      OpBuilder::InsertionGuard guard(builder);
+      builder.setInsertionPointToStart(&newOp.getRegion().front());
+      auto maketptrOp = state.createTTSMakeTensorPtrOp(builder, op.getLoc());
+      map.map(key, maketptrOp.getResult());
+    } else {
+      OpBuilder::InsertionGuard guard(builder);
+      builder.setInsertionPointToStart(&newOp.getRegion().front());
+
+      auto offset = state.scalar;
+      if (offset.getType().isa<IndexType>()) {
+        offset = builder.create<arith::IndexCastOp>(
+            op.getLoc(), builder.getI32Type(), offset);
+      }
+      auto addptrOp = builder.create<triton::AddPtrOp>(
+          op.getLoc(), state.source.getType(), state.source, offset);
+      map.map(key, addptrOp.getResult());
+    }
```

Comment:
similar comment, this logic of handling `state.rank == 0` is new, would be helpful if we have some comments and example snippet here

Reply:
Added comment above this section of code. Your comment actually made me realize that we don't need to create new `tt.addptr` at the beginning of the loop body, only `tts.make_tptr`. I removed dead code and clarified in the comment added.

Reply:
Closing for now. Please feel free to reopen.